ethgraphql/main.go

```diff
@@ -0,0 +1,1103 @@
+	block, err := b.resolve(ctx)
+	if err != nil || block == nil {
+		return nil, err
+	}
+	count := int32(len(block.Uncles()))
+	return &count, err
+}
+
+func (b *Block) Ommers(ctx context.Context) (*[]*Block, error) {
+	block, err := b.resolve(ctx)
+	if err != nil || block == nil {
+		return nil, err
+	}
+
+	ret := make([]*Block, 0, len(block.Uncles()))
+	for _, uncle := range block.Uncles() {
+		blockNumber := rpc.BlockNumber(uncle.Number.Uint64())
+		ret = append(ret, &Block{
+			backend: b.backend,
+			num:     &blockNumber,
+			hash:    uncle.Hash(),
+			header:  uncle,
+		})
+	}
+	return &ret, nil
+}
+
+func (b *Block) ExtraData(ctx context.Context) (hexutil.Bytes, error) {
+	header, err := b.resolveHeader(ctx)
+	if err != nil {
+		return hexutil.Bytes{}, err
+	}
+	return hexutil.Bytes(header.Extra), nil
+}
+
+func (b *Block) LogsBloom(ctx context.Context) (hexutil.Bytes, error) {
+	header, err := b.resolveHeader(ctx)
+	if err != nil {
+		return hexutil.Bytes{}, err
+	}
+	return hexutil.Bytes(header.Bloom.Bytes()), nil
+}
+
+func (b *Block) TotalDifficulty(ctx context.Context) (hexutil.Big, error) {
+	h := b.hash
+	if h == (common.Hash{}) {
+		header, err := b.resolveHeader(ctx)
+		if err != nil {
+			return hexutil.Big{}, err
+		}
+		h = header.Hash()
+	}
+
+	return hexutil.Big(*b.backend.GetTd(h)), nil
+}
+
+// BlockNumberArgs encapsulates arguments to accessors that specify a block number.
+type BlockNumberArgs struct {
+	Block *hexutil.Uint64
+}
+
+// Number returns the provided block number, or rpc.LatestBlockNumber if none
+// was provided.
+func (a BlockNumberArgs) Number() rpc.BlockNumber {
+	if a.Block != nil {
+		return rpc.BlockNumber(*a.Block)
+	}
+	return rpc.LatestBlockNumber
+}
+
+func (b *Block) Miner(ctx context.Context, args BlockNumberArgs) (*Account, error) {
+	block, err := b.resolve(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Account{
+		backend:     b.backend,
+		address:     block.Coinbase(),
+		blockNumber: args.Number(),
+	}, nil
+}
+
+func (b *Block) TransactionCount(ctx context.Context) (*int32, error) {
+	block, err := b.resolve(ctx)
+	if err != nil || block == nil {
+		return nil, err
+	}
+	count := int32(len(block.Transactions()))
+	return &count, err
+}
+
+func (b *Block) Transactions(ctx context.Context) (*[]*Transaction, error) {
+	block, err := b.resolve(ctx)
+	if err != nil || block == nil {
+		return nil, err
+	}
+
+	ret := make([]*Transaction, 0, len(block.Transactions()))
+	for i, tx := range block.Transactions() {
+		ret = append(ret, &Transaction{
+			backend: b.backend,
+			hash:    tx.Hash(),
+			tx:      tx,
+			block:   b,
+			index:   uint64(i),
+		})
+	}
+	return &ret, nil
+}
+
+func (b *Block) TransactionAt(ctx context.Context, args struct{ Index int32 }) (*Transaction, error) {
+	block, err := b.resolve(ctx)
+	if err != nil || block == nil {
+		return nil, err
+	}
+
+	txes := block.Transactions()
+	if args.Index < 0 || int(args.Index) >= len(txes) {
+		return nil, nil
```

Comment:
```suggestion
		return nil, errors.New("tx index out of bounds")
```

Reply:
The problem here is that callers may request an index that they don't know is out of bounds. For instance, to get the first uncle of every block in a range:
```
{
  block(from: 1000000, to: 2000000) {
    ommerAt(index: 0) {
      hash
    }
  }
}
```

Producing an error for each block without uncles would be unhelpful. I thought it would be more useful to return nil (and document that fact).